tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy

```diff
@@ -113,127 +77,95 @@ abstract class NimbusPluginExtension {
-                    fetchNimbusBinaries(project)
-                } else {
-                    println("Using local application services")
-                }
+            generateTask.configure {
+                fmlBinary = assembleToolsTask.flatMap { it.fmlBinary }
+                dependsOn validateTask
             }
-        }
-    }
 
-    // Try one or more hosts to download the given file.
-    // Return the hostname that successfully downloaded, or null if none succeeded.
-    static def tryDownload(File directory, String filename, String[] urlPrefixes) {
-        return urlPrefixes.find { prefix ->
-            def urlString = filename == null ? prefix : "$prefix/$filename"
-            try {
-                new URL(urlString).withInputStream { from ->
-                    new File(directory, filename).withOutputStream { out ->
-                        out << from;
-                    }
-                }
-                true
-            } catch (e) {
-                false
+            project.tasks.named {
+                it.contains('compile')
+            }.configureEach { task ->
+                task.dependsOn generateTask
             }
         }
     }
 
-    // Fetches and extracts the pre-built nimbus-fml binaries
-    def fetchNimbusBinaries(Project project) {
-        def asVersion = getProjectVersion()
-
-        def fmlPath = getFMLFile(project, asVersion)
-        println("Checking fml binaries in $fmlPath")
-        if (fmlPath.exists()) {
-            println("nimbus-fml already exists at $fmlPath")
-            return
-        }
+    def setupAssembleNimbusTools(Project project) {
+        return project.tasks.register('assembleNimbusTools', NimbusAssembleToolsTask) { task ->
+            group "Nimbus"
+            description "Fetch the Nimbus FML tools from Application Services"
 
-        def rootDirectory = getFMLRoot(project, asVersion)
-        def archive = new File(rootDirectory, "nimbus-fml.zip")
-        ensureDirExists(rootDirectory)
+            def asVersion = getProjectVersion()
+            def fmlRoot = getFMLRoot(project, asVersion)
 
-        if (!archive.exists()) {
-            println("Downloading nimbus-fml cli version $asVersion")
+            archiveFile = fmlRoot.map { it.file('nimbus-fml.zip') }
+            hashFile = fmlRoot.map { it.file('nimbus-fml.sha256') }
+            fmlBinary = fmlRoot.map { it.file(getFMLFile()) }
 
-            def successfulHost = tryDownload(archive.getParentFile(), archive.getName(),
+            fetch {
                 // Try archive.mozilla.org release first
-                "https://archive.mozilla.org/pub/app-services/releases/$asVersion",
-                // Try a github release next (TODO: remove this once we verify that publishing to
-                // archive.mozilla.org is working).
-                "https://github.com/mozilla/application-services/releases/download/v$asVersion",
```

Comment:
We haven't published `nimbus-fml.zip` artifacts to Application Services [since version 116](https://github.com/mozilla/application-services/releases/tag/v116.0), so I removed this.